### PR TITLE
added tree view and integrated it into component playground

### PR DESCRIPTION
**Pull request by amrdraz**
This pull requests settles issue #251 

currently I'm writing tests for the FixtureList component, but I would like some feedback on the current design

there were things I could have done more like remembering the state of each node in the tree propr to search but after trying it out it felt unnecessary 

@skidding  please review ux and any design recomendations

as requested instead of using a tree-view with lots of dependencies I built a simple light tree view and integrated under the FixtureList component

### CODE REVIEW

**Comment by skidding** [comment]
First of all, playing with this is fun! Good job getting this to work!

Here are my thoughts:

Is `/components/ComponentA/components/ComponentB/` a known pattern? I find it repetitive because we keep adding "components" dirs. Doesn't look that good in the the playground either. How about putting the Tree component at the same level with FixtureList? Or is it possible to get rid of the "components" altogether and have `components/ComponentA/ComponentB`? I'm starting to see the value of nesting them, especially in the tree view...

I couldn't read all the code yet, so I'll have to come back to it. I like that you didn't introduce a bloated lib and kept it minimal, but it's quite a lot of code for me to parse so can you please write a small description of the new abstractions (components, classes, etc) you added and what each does? It will be very useful in guiding me when reading the code (but also useful to add as documentation via comments).

One thing I noticed is that you keep derived data into state, like the `filteredFixtures` and `selectedPath`, which is an anti pattern. A core tenet of React is to keep state minimal, and anything that can be derived should be computed on render. If it ever becomes an optimization problem we can memoize. So, unless you have a reason I'm not aware of I'd like to go back to using [this pattern](https://facebook.github.io/react/docs/thinking-in-react.html#step-3-identify-the-minimal-but-complete-representation-of-ui-state).

Regarding the design. Can we use the existing colors, spacings and hover please? The folder icon as well, it looks friendlier.
image

The new design is nice to play with, but has a bit of Win98 feel to it :)
image

I know the tree view changes the markup, but here's an example from Atom where the selection bg goes all the way from left to right even in a nested structure:
image
Let's use the same.

The yellow `js` icons looks cool on the Cosmos blue, but it gets redundant after a while. In Atom you have many types of files, so it makes more sense (although I hid the icons anyway), but here *all* non-folder items are fixtures, so I say we keep them clean.

Let me know if there's anything I can help with. I'll probably add some inline comments as well after this.

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "xo.enable": true
```
Comment (skidding) [inline .vscode/settings.json]: No editor specific config files please. `.vscode` can go in your global gitignore.

File: packages/react-cosmos-component-playground/package.json [hunk 2]
```diff
@@ -8,6 +8,9 @@
     "lib"
   ],
   "main": "lib/index.js",
+  "scripts": {
+    "cosmos": "NODE_ENV=development cosmos"
```
Comment (skidding) [inline packages/react-cosmos-component-playground/package.json]: No need for this, you can already run `yarn run start:playground` from the root project dir.

File: packages/react-cosmos-component-playground/package.json [hunk 3]
```diff
@@ -21,8 +24,10 @@
     "react": "^15.5.4",
     "react-cosmos-loader": "^2.0.0-beta.24",
     "react-cosmos-state-proxy": "^2.0.0-beta.24",
+    "react-cosmos-webpack": "^2.0.0-beta.25",
     "react-dom": "^15.5.4",
-    "react-querystring-router": "^2.0.0-beta.23"
+    "react-querystring-router": "^2.0.0-beta.23",
+    "react-test-renderer": "^15.6.1"
```
Comment (skidding) [inline packages/react-cosmos-component-playground/package.json]: No need for this, test dependencies are installed globally in the root package.json to not have to add them in every package.

**Comment by amrdraz** [comment]
`/components/ComponentA/components/ComponentB/` is pattern I picked up [from this post](https://medium.com/@alexmngn/how-to-better-organize-your-react-applications-2fd3ea1920f1). But I agree that while it helps keep things clean when I have other type of files that are specific to the component (reducer, actions, services etc..), it gets repetitive but. I'm not too attached to it, plus I try to conform to the projects conventions, I mainly kept it to keep a deeper tree when testing. (was even thinking of making the tree ignore it XD but then that would have invited all sorts of trouble)

----

The code is mostly tests actually there are 3 files that were introduced and a modification to the FixtureList to adapt the change.

We have the `Tree` component and its `TreeService` and we have a subcomponent of Tree called `Node` which renders each node in the tree. 

**Node**

The Node component takes props

```js
{
  name: string, // the name that appears next to the folder
  path: string, // the path up until this point in the tree used in title and to trigger actions on this node
 open: boolean, // if this node is not a lead then if true it will show the children and the down arrow
 active: boolean, // is true when the node is selected
 isLeaf: boolean, // is true when node path matches the path used to construct this node and the path ends with a value
 value: any, // a value that is set for the leaf at the end of the path can be anything
 onLeafClick: (node)=>{}, // a function that is fired when a leaf is clicked
 onNodeClick: (node)=>{}, // a function that is fired when a non leaf is clicked
 treeService: TreeService, // an instance of the TreeService used to select and toggleOpen
}
```
instance methods

```js
toggleOpen() // opens the node ats props.path
selectNode() // selects the node at props.path
```

**Tree**

initializes the TreeService and subscribes to it's state
updates the TreeService with changes occur to props that may change the state

can take the following props
```js
{
  fixtures: object //  an object map of a path to the components with and array of fixture names used
  paths: object // a processed paths object with path to fixture and value to be set on the leaf
  tree: object // an initial tree state
  selectedPath: string // path to be selected after the tree has been constructed
  onLeafClick: (node)=>{} // a function that is fired when a leaf is clicked passed the node as argument
}
```

When rendering the tree will sort leafs at the beginning in order to place the fixtures right under the folder then subcomponents can show

**TreeService**

This is where the tree's state is managed, it can be initialized with a tree state, a paths object, or a fixtures object

The tree is constructed using a paths object of the form `{ [path]: value }`
for each path in paths, a node is created for every fragment of the path
The node with the last fragment in the path has the value attached to it.

all nodes are initially open

> a path fragment array is the results of splitting the path on `/` (with this be a cross platform issue?)

if initialized with fixtures, fixtures are converted to the paths object form

```js
{
  `componentA`: [`fixture1`, `fixture2`],
  `componentA/B`: [`fixture1`],
  `componentC`: []
}
```
becomes
```
{
  `componentA/fixture1`: {
     component: 'componentA',
     fixture: 'fixture1'
   },
  `componentA/fixture2`: {
     component: 'componentA',
     fixture: 'fixture2'
   },
  `componentA/B/fixture1`: {
     component: 'componentA/B',
     fixture: 'fixture1'
   },
   `componentA/B`:  undefined
}
```

the tree is immutable and always updates the path the changed

**TreeService.select(path, multi)**

set the node at the end of the path to `active:true`

multi is a toggle option for multi path select, it is not needed, I was mostly playing around :P 

**TreeService.toggleOpen(path)**

toggle the `open` property of the node at the end of the path

I will update the files to reflect documentation
------

regarding the state upon reflection I agree there is no need to retain the fixture or selectedPath in the state I will move them back to be computed in render.

-----

I thought of a means to highlight the background and can change the folder icon
I personally prefer having something to anchor my eye on regarding the fixture icon, but I'm not attached to it.
I do not agree completely with regards to the spacing between elements, the font felt a bit large when in the tree view.

Will apply updates on Monday night I think

**Comment by skidding** [comment]
> /components/ComponentA/components/ComponentB/ is pattern I picked up from this post.

Thanks for sharing the post. I read it and it I like it. And thanks for being flexible. I prefer using a flat structure in the CP for now to focus on the tree view functionality itself in this PR. You can open a subsequent PR to convert the entire CP package structure to the component/service pattern and link to the article in the contributing guide.

CP doesn't have many files, but I'm more interested in how well #440 works with this pattern. It would make a great case study. Maybe even add a section to README/Usage, saying how the config should look like if you use that pattern, similar to _Integration with popular tools_. BUT, let's focus on this PR.

>  I mainly kept it to keep a deeper tree when testing.

Makes perfect sense! Feel free to keep it like this until we finish the review process.

> Node

- Given the above, how about naming this `TreeNode` for now?
- `active: boolean`, how about `selected`? It's more suggestive
- I only see `onNodeClick` passed in tests. If you inspect the components you see no node receives `onNodeClick` – Is this WIP?
  image

Besides components and the tree class, it seems like we introduce _types_ so let's define them. In the future it will allow us to use something like Flow or TS, but for now let's document them to organize data and keep a clear image in our heads.

Like you said, we start from `fixtures`:

```js
{
  `componentA`: [`fixture1`, `fixture2`],
  `componentA/B`: [`fixture1`],
  `componentC`: []
}
```

Then we create an intermediary `paths` object:

```js
{
  `componentA/fixture1`: {
     component: 'componentA',
     fixture: 'fixture1'
   },
  `componentA/fixture2`: {
     component: 'componentA',
     fixture: 'fixture2'
   },
  `componentA/B/fixture1`: {
     component: 'componentA/B',
     fixture: 'fixture1'
   },
   `componentA/C`:  undefined
}
```

We need to take #383 into account. If we don't prepare for it, we'll have to refactor soon. In the future we'll be able to select components as well. So we'll have non-leaf nodes that are also selectable. Maybe this?

```js
{
  `componentA/B`: {
    component: 'componentA/B'
  },
  `componentA/fixture1`: {
    component: 'componentA',
    fixture: 'fixture1'
  },
  `componentA/fixture2`: {
    component: 'componentA',
    fixture: 'fixture2'
  },
  `componentA/B`: {
    component: 'componentA/B'
  },
  `componentA/B/fixture1`: {
    component: 'componentA/B',
    fixture: 'fixture1'
  },
  `componentA/C`: {
    component: 'componentA/C'
  }
}
```

Then we have the `TreeNode` type, which I think is the most important one. We must prevent inconsistent and redundant data.
image
For example we might know a node is a leaf by checking if it has children. Please correct me if I'm wrong.

Proposal:
```js
// TreeNode type
{
  children: [] // list of TreeNodes, always Array
  open: true, // always boolean
  selected: false, // always boolean
  name: 'components',
  value: {}
}
// isLeaf = children.length > 0
```

There's a lot of state duplication. The Tree service stores paths and the tree, but then you also keep paths and the tree in the Tree component, and sync the two using event listeners. We can get rid of so much code! ⚔️ 

So here's my biggest ask for this review: Can you transform the Tree service from a class with event listeners into a functional service? I'm positive it's possible. I see one big function: `createTreeFromPaths(paths, selectedPath)` (or from `fixtures` directly, which ever is easier – it's not even clear yet to my why we need the intermediary `paths` but it probably makes sense to keep the tree service cleaner). This one returns a tree but you never really have to add it to state, as it can always be derived deterministically. Then you have `selectedPath`, which also doesn't need to be added to state, because you can derive it from `urlParams.component` and `urlParams.fixture`. So whenever a node is clicked, the event bubbles up to FixtureLoader, which updates urlParams from selected path, and then that will trickled down (Flux style), effectively rendering a different `selectedPath` and `tree` respectively.  What do you think, doable?

Thank you @amrdraz!

**Comment by skidding** [comment]
Closing this in favor of #448 

Thanks @amrdraz!

**Pull request closed without merging**